doc/ConnectionPooling.md

```diff
@@ -0,0 +1,387 @@
+## Using Connection Pools
+
+### Multiple Connection Pools
+
+Snowflake .NET Driver v4.0.0 provides multiple pools with couple of additional features in comparison to the previous implementation.
+Each pool is identified by the entire <b>connection string</b>. Order of connection string parameters is significant and the same connection parameters
+ordered differently lead to two different pools being used.
+
+All the pool parameters can be controlled from the connection string.
+
+Pool interface is also maintained by the [SnowflakeDbConnectionPool.cs](https://github.com/snowflakedb/snowflake-connector-net/blob/master/Snowflake.Data/Client/SnowflakeDbConnectionPool.cs).
+However, some operations (eg. setting pool parameters from this SnowflakeDbConnectionPool class) are not possible having in mind multiple pools and possibly their different setup.
+For that a [SnowflakeDbSessionPool.cs](/Snowflake.Data/Client/SnowflakeDbSessionPool.cs) is provided by
+- [SnowflakeDbSessionPool.GetPool(connectionString)](https://github.com/snowflakedb/snowflake-connector-net/blob/master/Snowflake.Data/Client/SnowflakeDbConnectionPool.cs#L45)
+- [SnowflakeDbSessionPool.GetPool(connectionString, securePassword)](https://github.com/snowflakedb/snowflake-connector-net/blob/master/Snowflake.Data/Client/SnowflakeDbConnectionPool.cs#L51).
+to control pool settings from the code. Changed pool settings are not reflected by their connection string therefore recommended way is to control the pool from the connection string.
+
+### Pool Lifecycle
+
+Single pool is instantiated each time an application creates and opens a connection for the first time using particular connection string.
+From that moment the pool tracks and maintains connections matching exactly this connection string.
+Pool is responsible for destroying and recreating connections which are old enough (see [Expiration Timeout](#expiration-timeout)).
+Number of connections is maintained within [Minimum pool size](#minimum-pool-size) and [Maximum pool size](#maximum-pool-size).
+Connections in all their statuses are tracked:
+- opening phase
+- busy phase
+- closed and returned to the pool (idle)
+User can clean up the pool using methods: [Clear Pool](#clear-pool).
+
+### Connection Lifecycle
+
+#### Opening
+
+Before connection is opened it is `scheduled` to be opened in the pool on a waiting queue. Connections scheduled to open are counted to the pool size.
+After connection is successfully opened it is counted as a `busy` in a separate list. Pool can provide already opened `idle` connection or if there are none it will schedule new one to open.
```

Comment:
maybe: "After connection is successfully opened and returned to the user it is regarded as busy" - I would add that we are talking about returning it to the user, because if we do it to ensure minPoolSize it would go to idle list. And I would not write about "lists" because it is a technical detail.